New: communication back from custom import script

JeWe37 opened the pull request:
#### Database Migration
NO

#### Description
There are a few cases that were not handled for custom import scripts yet:

* Custom import scripts that generate additional extra files to be imported.
* Custom import scripts that remux the original media file to a different container, such that its extension must change
* Wanting to import extra files via a custom import script(as deferred to a later date in the original #5538)

All of these use cases are addressed as follows: A custom import script can now communicate back to Sonarr by means of a temporary file, the path to which is passed as an environment variable. This avoids parsing a potentially polluted stdout, however I am open to changes in this respect. This communication is used in order to inform Sonarr of: The potentially changed name of the media file, and which extras have been imported, or equivalently generated, by the custom import script.

Still a WIP as I haven't really tested this yet and am just leaving it for initial feedback.

#### Todos
- [ ] Tests
- [ ] Wiki Updates


#### Issues Fixed or Closed by this PR

*

markus101 reviewed (changes requested):
Left some inline comments.

JeWe37 commented:
Ultimately I am not horribly invested in the file idea. I'll switch to just looking at stdout(and fix the other things).

JeWe37 commented:
Switched to using stdout and addressed the other comments. In testing it turned out there was a dependency loop, broke that by adding some more information to the series scanned event which I think makes sense architecturally.

markus101 commented on src/NzbDrone.Core/MediaFiles/ScriptImportDecider.cs:
```diff
@@ -25,23 +28,75 @@ public class ImportScriptService : IImportScript
+                }
+            }
+
+            return new ScriptImportInfo(possibleExtraFiles, mediaFile ?? defaultMediaFile);
```
It shouldn't be the responsibility of the output parser to return a default.

markus101 commented on src/NzbDrone.Core/MediaFiles/ScriptImportDecider.cs:
```diff
@@ -25,23 +28,75 @@ public class ImportScriptService : IImportScript
+
+        private static readonly Regex ExtraFileLine = new Regex(@"^\[SonarrExtraFile\]\s*(?<extraFileName>.*)$", RegexOptions.Compiled);
+
+        private ScriptImportInfo ProcessStdout(List<ProcessOutputLine> processOutputLines, string defaultMediaFile)
```
`ProcessStdout` -> `ProcessOutput`

markus101 commented on src/NzbDrone.Core/MediaFiles/ScriptImportDecider.cs:
```diff
@@ -118,12 +173,27 @@ public ScriptImportDecision TryImport(string sourcePath, string destinationFileP
+            episodeFile.RelativePath = series.Path.GetRelativePath(mediaFile);
+            episodeFile.Path = mediaFile;
+
+            if ((processOutput.ExitCode & 0x4) == 0x4)
+            {
+                localEpisode.ShouldImportExtras = true;
+            }
+
+            switch (processOutput.ExitCode & 0x3)
```
Should be able to achieve all this without mixing in the bit-wise operator along with an if and a switch which will make it easier to read and maintain.

JeWe37 replied:
Personally I don't really like my new alternative any better...

markus101 replied:
Did you forget to push these changes up?

JeWe37 replied:
I did indeed...... Fixed.

markus101 commented on src/NzbDrone.Core/MediaFiles/ScriptImportDecider.cs:
```diff
@@ -25,23 +28,75 @@ public class ImportScriptService : IImportScript
             _logger = logger;
         }
 
+        private static readonly Regex MediaFileLine = new Regex(@"^\[SonarrMediaFile\]\s*(?<mediaFileName>.*)$", RegexOptions.Compiled);
+
+        private static readonly Regex ExtraFileLine = new Regex(@"^\[SonarrExtraFile\]\s*(?<extraFileName>.*)$", RegexOptions.Compiled);
```
I think something like:

```suggestion
        private static readonly Regex OutputRegex = new Regex(@"^\[(?:(?<mediaFile>MediaFile)|(?<extraFile>ExtraFile))\]\s(?<fileName>.+)$", RegexOptions.Compiled);
```

for both these regexes to be a little more strict on the format and ensuring the script provided information instead of a null/empty value.

Changed to `\[MediaFile\]` to make it universal with downstream apps and merged into a single regex.

markus101 commented on src/NzbDrone.Core/MediaFiles/ScriptImportDecider.cs:
```diff
@@ -25,23 +28,75 @@ public class ImportScriptService : IImportScript
+
+            foreach (var line in processOutputLines)
+            {
+                if (MediaFileLine.Match(line.Content) is var match && match.Success)
```
With the suggestion to use one regex this will need to be reworked a bit, but should be pretty similar.

JeWe37 replied:
I did indeed...... Fixed.

markus101 reviewed (changes requested):
Left some ideas to simplify it a little bit, but looking good.

markus101 reviewed (changes requested):
Getting away from the exit code driving behaviour would be good, since this is still in a beta release a breaking change there would be acceptable.

JeWe37 commented on src/NzbDrone.Core/MediaFiles/ScriptImportDecider.cs:
```diff
@@ -25,23 +27,71 @@ public class ImportScriptService : IImportScript
+
+                    if (!_diskProvider.FileExists(lastAdded))
+                    {
+                        throw new ScriptImportException("Script output contains non-existent possible extra file: {0}", lastAdded);
```
Switched to emitting a warning. Import state should still be fine I think, I don't think anything was added to the database by this point? But changed anyway.

JeWe37 commented on src/NzbDrone.Core/MediaFiles/EpisodeImport/ImportApprovedEpisodes.cs:
```diff
@@ -130,7 +133,20 @@ public List<ImportResult> Import(List<ImportDecision> decisions, bool newDownloa
-                        _extraService.ImportEpisode(localEpisode, episodeFile, copyOnly);
+                        if (localEpisode.ScriptImported)
+                        {
+                            _existingExtraFiles.ImportFileList(localEpisode.Series, localEpisode.PossibleExtraFiles);
```
This is actually intended, "ShouldImportExtras" only is supposed to instruct sonarr to import extras. Even if we prevent Sonarr from importing in the script, I think it still makes sense to be able to import extras via the script ourselves. This is mostly just for the case where the script doesn't handle extras, but we want to still import them via Sonarr's stock logic.

markus101 commented on src/NzbDrone.Core/Extras/ExistingExtraFileService.cs:
```diff
@@ -50,6 +38,17 @@ public void Handle(SeriesScannedEvent message)
+
+        public void Handle(SeriesScannedEvent message)
+        {
+            var series = message.Series;
+
+            var possibleExtraFiles = message.PossibleExtraFiles;
+
+            var importedFiles = ImportExtraFiles(series, possibleExtraFiles);
```
```suggestion
            var series = message.Series;
            var possibleExtraFiles = message.PossibleExtraFiles;
            var importedFiles = ImportExtraFiles(series, possibleExtraFiles);
```
